src/util/mmap_impl_win32.h

```diff
@@ -0,0 +1,241 @@
+  }
+  else {
+      LARGE_INTEGER size;
+      if (GetFileSizeEx(fd, &size)) {
+          //size.QuadPart is int64
+          this->size_ = static_cast<size_t>(size.QuadPart);
+      }
+      else {
+        return Status::InvalidState()
+               << "unable to get size of file: " << filename
+               << " errcode=" << GetLastError();
+      }
+  }
+
+  this->fd_ = fd;
+
+  return Status::Ok();
+}
+
+inline static DWORD get_system_granuality() {
+  SYSTEM_INFO info;
+  GetSystemInfo(&info);
+  return info.dwAllocationGranularity;
+}
+
+inline static bool set_file_len(void *handle, size_t size) {
+    LARGE_INTEGER info_size;
+    info_size.QuadPart = static_cast<__int64>(size);
+    FILE_END_OF_FILE_INFO info = {info_size};
+
+    //Makes minimum supported platform Windows Vista
+    return static_cast<bool>(SetFileInformationByHandle(handle,
+                                                        FileEndOfFileInfo,
+                                                        &info,
+                                                        sizeof(info)));
+}
+
+Status MappedFile::map(MappedFileFragment *view, size_t offset, size_t size,
+                       bool useHuge) {
+  if (!view->isClean()) {
+    return JPPS_INVALID_STATE << "fragment for the file " << filename_
+                              << " is already mapped";
+  }
+
+  DWORD protection;
+  DWORD map_access;
+  const auto file_end = offset + size;
+
+  switch (type_) {
+    case MMapType::ReadOnly:
+      if (file_end > size_) {
+        return JPPS_INVALID_PARAMETER
+               << "offset=" << offset << " and size=" << size
+               << " point after the end of file " << filename_
+               << " of size=" << size_;
+      }
+      map_access = FILE_MAP_READ;
+      protection = PAGE_READONLY;
+      break;
+    case MMapType::ReadWrite:
+      if (file_end > size_) {
+        if (!set_file_len(this->fd_, file_end)) {
+          return JPPS_INVALID_STATE << "could not extend file "
+                                    << filename_ << " to " << file_end
+                                    << "bytes";
+        }
+      }
+      map_access = FILE_MAP_WRITE; //read/write access
+      protection = PAGE_READWRITE;
+      break;
+  }
+
+  const auto mapping = CreateFileMappingW(this->fd_, nullptr, protection,
+                                          0, 0, nullptr);
+  if (mapping == NULL) {
+    return JPPS_INVALID_STATE << "Cannot open file mapping for "
+                              << filename_
+                              << ". Error: " << GetLastError();
+  }
+
+  const DWORD granuality = get_system_granuality();
+
+  const size_t alignment = offset % static_cast<size_t>(get_system_granuality());
+  const size_t aligned_offset = offset - alignment;
+  const size_t aligned_len = size + alignment;
+
+  const auto map_view = MapViewOfFile(mapping, map_access,
+                                      static_cast<DWORD>(aligned_offset >> 16 >> 16),
```

Comment:
`>> 32`?

Reply:
Ah yes, I was using some used by me library as reference for mmap